File: rust/src/day1.rs
use std::fs::File;
use std::io::{BufReader, BufRead};

fn fuel_required(mass : i32) -> i32 {
    mass / 3 - 2
}

fn total_fuel_required(mass : i32) -> i32 {     
    match fuel_required(mass) {
        -2 | -1 | 0 => 0,
        other => other + total_fuel_required(other)
    }        
}

pub fn run() {
    println!("--- Day 1: The Tyranny of the Rocket Equation ---");
    let modules : Vec<i32> = 
        BufReader::new(
            File::open("resources/day1input.txt").unwrap())
            .lines()
            .map(|il| il.unwrap().parse().unwrap())
            .collect();

    let mut sum : i32 = 0;
    for module in &modules {
        sum += fuel_required(*module);
    }
    println!("Fuel required for modules: {}", sum);

    sum = 0;
    for module in &modules {
        sum += total_fuel_required(*module);
    }
    println!("Total fuel required: {}", sum);
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! fuel_required_tests {
        ($($name:ident: $value:expr,)*) => {
        $(
            #[test]
            fn $name() {
                let (input, expected) = $value;
                assert_eq!(expected, fuel_required(input));
            }
        )*
        }
    }

    fuel_required_tests! { 
        fuel_required_12: (12, 2),
        fuel_required_14: (14, 2),
        fuel_required_1969: (1969, 654),
        fuel_required_100756: (100756, 33583),
    }


    macro_rules! total_fuel_required_tests { ($($name:ident: $value:expr,)*) => { $(
    
    #[test]
    fn $name() {
        let (input, expected) = $value;
        assert_eq!(expected, total_fuel_required(input));
    }

    )*}}

    total_fuel_required_tests! {         
        total_fuel_required_14: (14, 2),
        total_fuel_required_1969: (1969, 966),
        total_fuel_required_100756: (100756, 50346),
    }
}

File: lua/day1.lua
require "test"
require "util"

local function sum(table, scoreOf)
    local sum = 0
    for key, value in pairs(table) do
        sum = sum + scoreOf(value)
    end
    return sum
end

local function fuelRequired(mass)
    return math.floor(mass / 3) - 2
end

local function moduleFuelRequirement(modules)
    return sum(modules, fuelRequired)
end

local function totalFuelRequired(mass)
    local required = fuelRequired(mass)
    if required <= 0 then
        return 0
    else
        return required + totalFuelRequired(required)
    end
end

local function totalModuleFuelRequirement(modules)
    return sum(modules, totalFuelRequired)
end

local function test()
    doTests{
        ["Base fuel requirement"] = {
            { { 12, 2 }, { 14, 2 }, {1969, 654 }, { 100756, 33583 } },
            function (_, mass, fuel) 
                assert(fuelRequired(mass) == fuel)                
            end
        },
        ["Total fuel requirement"] = {
            { { 14, 2 }, { 1969, 966 }, { 100756, 50346 } },
            function(_, mass, fuel) 
                assert(totalFuelRequired(mass) == fuel)
            end
        }
    }
end

local modules = {}

local function load()
    for line in readLines("../resources/day1/input.txt") do
        table.insert(modules, line)
    end
end

local function runDay1()    
    print(moduleFuelRequirement(modules))
end

local function runDay2()
    print(totalModuleFuelRequirement(modules))
end

return {
    ["title"] = "The Tyranny of the Rocket Equation",
    ["tests"] = test,
    ["pre"] = load,
    ["part1"] = runDay1,
    ["part2"] = runDay2,
}